keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,546 @@
+
+        // Status defines the current state of a hook.
+        // +optional
+        Status ExecutionHookStatus
+}
+```
+
+Here is the definition of the ExecutionHookSpec:
+
+```
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Snapshot Controller.
+type ExecutionHookSpec struct {
+        // PodSelection defines how to select pods and containers to run
+	// the executionhook. This field is required.
+        PodSelection PodSelection
+
+        // Name of the HookAction. This is required.
+        HookActionName string
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where one of them must be defined so that the hook controller knows where to
+// run the hook.
+type PodSelection struct {
+        // PodContainerNamesList lists the pods/containers on which the ExecutionHook
+        // should be executed. If not specified, the ExecutionHook controller will find
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *PodContainerSelector
+}
+
+type PodContainerNames struct {
+        PodName string
+        ContainerNames []string
+}
+
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // ContainerSelector is a selector which must be true for the hook to run on the container
+        // Hook controller will find all containers in the selected pods based on ContainerSelector,
+        // if specified.
+        // The ContainerSeletor maps pod label to container name(s) so that the controller only select
+	// the containers that are listed in the map. Otherwise, all containers of the pods that match
+	// the pod label selector will be selected.
+        // +optional
+        ContainerSelector map[string]string
+ }
+```
+
+Here is the definition of the ExecutionHookStatus. This represents the current state of a hook for all selected containers in the selected pods.
+
+```
+// ExecutionHookStatus represents the current state of a hook
+type ExecutionHookStatus struct {
+        // This is a list of ContainerExecutionHookStatus, with each status representing information
+        // about how hook is executed in a container, including pod name, container name, ActionTimestamp,
+        // ActionSucceed, etc.
+        // +optional
+        ContainerExecutionHookStatuses []ContainerExecutionHookStatus
+
+        // Action Summary status
```

Comment:
The idea is that this summary status is set to true only if actions are executed successfully in all selected containers. If any container fails to execute this action or times out, this will be set to false. 
If using action completed, then it will set to true after all containers finish executing (either succeed or fail or timeout).
I think actionSucceeded can make it easy for user to check the succeeded state instead of going through all detailed status. It will also indicate failure as early as possible.

Reply:
if a user wants to know as early as possible, they are watching.  If they are watching, they will see the failure of the individual pod

Reply:
Let's remove this since it is not strictly necessary (and a little confusing). We can reintroduce in the future if necessary).

Reply:
Sure.  Removed.